employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/SkillService.java

```diff
@@ -0,0 +1,116 @@
+
+    public SkillService(SkillRepository skillRepository) {
+        this.skillRepository = skillRepository;
+    }
+
+    @Transactional
+    public List<Skill> getSkillList(Integer tenantId) {
+        return skillRepository.findAllByTenantId(tenantId);
+    }
+
+    @Transactional
+    public Skill getSkill(Integer tenantId, Long id) {
+        Optional<Skill> skillOptional = skillRepository.findById(id);
+
+        if (!skillOptional.isPresent()) {
+            throw new EntityNotFoundException("No Skill entity found with ID (" + id + ").");
+        }
+
+        validateTenantIdParameter(tenantId, skillOptional.get());
+        return skillRepository.findById(id).get();
+    }
+
+    @Transactional
+    public Boolean deleteSkill(Integer tenantId, Long id) {
+        Optional<Skill> skillOptional = skillRepository.findById(id);
+
+        if (!skillOptional.isPresent()) {
+            throw new EntityNotFoundException("No Skill entity found with ID (" + id + ").");
+        }
+
+        validateTenantIdParameter(tenantId, skillOptional.get());
+        skillRepository.deleteById(id);
+        return true;
+    }
+
+    @Transactional
+    public Skill createSkill(Integer tenantId, Skill skill) {
+        validateTenantIdParameter(tenantId, skill);
+
+        String name = skill.getName();
+
+        //If name is null
+        if (name == null) {
+            throw new IllegalStateException("Skill entity name with tenantId (" + tenantId + ") cannot be null.");
+        }
+
+        List<Skill> skillList = skillRepository.findAllByTenantId(tenantId);
+
+        for (Skill s : skillList) {
+            //If skill already exists
+            if (s.getName().equals(name)) {
+                throw new EntityExistsException("Skill entity with tenantId (" + tenantId + ") and name (" + name
+                        + ") already exists.");
+            }
+        }
+
+        return skillRepository.save(skill);
+    }
+
+    @Transactional
+    public Skill updateSkill(Integer tenantId, Skill skill) {
+        validateTenantIdParameter(tenantId, skill);
+
+        Optional<Skill> skillOptional = skillRepository.findById(skill.getId());
+
+        //If name is null
+        if (skill.getName() == null) {
+            throw new IllegalStateException("Skill entity name with tenantId (" + tenantId + ") cannot be null.");
+        }
+        //Skill not found
+        else if (!skillOptional.isPresent()) {
+            throw new EntityNotFoundException("Skill entity with ID (" + skill.getId() + ") not found.");
+        }
+        //Skill changes tenantId
+        else if (!skillOptional.get().getTenantId().equals(skill.getTenantId())) {
+            throw new IllegalStateException("Skill entity with tenantId (" + skillOptional.get().getTenantId()
+                    + ") cannot change tenants.");
+        }
+
+        skillRepository.deleteById(skill.getId());
+        return skillRepository.save(skill);
```

Comment:
Don't delete the existing row when you want to update it. `reposiotry.save()` creates a new row for non-existing entity (doesn't have an id) and updates the row if the entity has an id (i.e. exists).

Reply:
Ah, so if `skill` has an id (ie. has been persisted) then `repository.save()` won't create a new row and will simply update? Good to know.